// THIS FILE IS AUTOMATICALLY GENERATED. DO NOT EDIT.
/**
The MIT License (MIT)

Copyright (c) 2017 Yuki Takei(noppoMan)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

import AWSSDKSwiftCore

/// Error enum for Clouddirectory
public enum ClouddirectoryError: AWSErrorType {
    case internalServiceException(message: String?)
    case invalidArnException(message: String?)
    case retryableConflictException(message: String?)
    case validationException(message: String?)
    case limitExceededException(message: String?)
    case accessDeniedException(message: String?)
    case resourceNotFoundException(message: String?)
    case schemaAlreadyPublishedException(message: String?)
    case directoryAlreadyExistsException(message: String?)
    case invalidAttachmentException(message: String?)
    case facetValidationException(message: String?)
    case directoryNotEnabledException(message: String?)
    case linkNameAlreadyInUseException(message: String?)
    case directoryDeletedException(message: String?)
    case facetNotFoundException(message: String?)
    case invalidNextTokenException(message: String?)
    case invalidTaggingRequestException(message: String?)
    case directoryNotDisabledException(message: String?)
    case objectNotDetachedException(message: String?)
    case facetAlreadyExistsException(message: String?)
    case invalidRuleException(message: String?)
    case invalidFacetUpdateException(message: String?)
    case schemaAlreadyExistsException(message: String?)
    case stillContainsLinksException(message: String?)
    case cannotListParentOfRootException(message: String?)
    case invalidSchemaDocException(message: String?)
    case notPolicyException(message: String?)
    case unsupportedIndexTypeException(message: String?)
    case batchWriteException(message: String?)
    case notIndexException(message: String?)
    case notNodeException(message: String?)
    case facetInUseException(message: String?)
    case objectAlreadyDetachedException(message: String?)
    case indexedAttributeMissingException(message: String?)
}

extension ClouddirectoryError {
    public init?(errorCode: String, message: String?){
        var errorCode = errorCode
        if let index = errorCode.index(of: "#") {
            errorCode = errorCode.substring(from: errorCode.index(index, offsetBy: 1))
        }
        switch errorCode {
        case "InternalServiceException":
            self = .internalServiceException(message: message)
        case "InvalidArnException":
            self = .invalidArnException(message: message)
        case "RetryableConflictException":
            self = .retryableConflictException(message: message)
        case "ValidationException":
            self = .validationException(message: message)
        case "LimitExceededException":
            self = .limitExceededException(message: message)
        case "AccessDeniedException":
            self = .accessDeniedException(message: message)
        case "ResourceNotFoundException":
            self = .resourceNotFoundException(message: message)
        case "SchemaAlreadyPublishedException":
            self = .schemaAlreadyPublishedException(message: message)
        case "DirectoryAlreadyExistsException":
            self = .directoryAlreadyExistsException(message: message)
        case "InvalidAttachmentException":
            self = .invalidAttachmentException(message: message)
        case "FacetValidationException":
            self = .facetValidationException(message: message)
        case "DirectoryNotEnabledException":
            self = .directoryNotEnabledException(message: message)
        case "LinkNameAlreadyInUseException":
            self = .linkNameAlreadyInUseException(message: message)
        case "DirectoryDeletedException":
            self = .directoryDeletedException(message: message)
        case "FacetNotFoundException":
            self = .facetNotFoundException(message: message)
        case "InvalidNextTokenException":
            self = .invalidNextTokenException(message: message)
        case "InvalidTaggingRequestException":
            self = .invalidTaggingRequestException(message: message)
        case "DirectoryNotDisabledException":
            self = .directoryNotDisabledException(message: message)
        case "ObjectNotDetachedException":
            self = .objectNotDetachedException(message: message)
        case "FacetAlreadyExistsException":
            self = .facetAlreadyExistsException(message: message)
        case "InvalidRuleException":
            self = .invalidRuleException(message: message)
        case "InvalidFacetUpdateException":
            self = .invalidFacetUpdateException(message: message)
        case "SchemaAlreadyExistsException":
            self = .schemaAlreadyExistsException(message: message)
        case "StillContainsLinksException":
            self = .stillContainsLinksException(message: message)
        case "CannotListParentOfRootException":
            self = .cannotListParentOfRootException(message: message)
        case "InvalidSchemaDocException":
            self = .invalidSchemaDocException(message: message)
        case "NotPolicyException":
            self = .notPolicyException(message: message)
        case "UnsupportedIndexTypeException":
            self = .unsupportedIndexTypeException(message: message)
        case "BatchWriteException":
            self = .batchWriteException(message: message)
        case "NotIndexException":
            self = .notIndexException(message: message)
        case "NotNodeException":
            self = .notNodeException(message: message)
        case "FacetInUseException":
            self = .facetInUseException(message: message)
        case "ObjectAlreadyDetachedException":
            self = .objectAlreadyDetachedException(message: message)
        case "IndexedAttributeMissingException":
            self = .indexedAttributeMissingException(message: message)
        default:
            return nil
        }
    }
}